File: game-server/src/main/java/work/fking/pangya/game/net/ClientPacketDispatcher.java
package work.fking.pangya.game.net;

import io.netty.channel.ChannelHandlerContext;
import io.netty.channel.SimpleChannelInboundHandler;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import work.fking.pangya.game.GameServer;
import work.fking.pangya.game.player.Player;

public class ClientPacketDispatcher extends SimpleChannelInboundHandler<ClientPacket> {

    private static final Logger LOGGER = LoggerFactory.getLogger(ClientPacketDispatcher.class);

    private final GameServer gameServer;
    private final Player player;
    private final ClientPacketHandler[] handlers;

    private ClientPacketDispatcher(GameServer gameServer, Player player, ClientPacketHandler[] handlers) {
        this.gameServer = gameServer;
        this.player = player;
        this.handlers = handlers;
    }

    public static ClientPacketDispatcher create(GameServer gameServer, Player player, ClientPacketHandler[] handlers) {
        return new ClientPacketDispatcher(gameServer, player, handlers);
    }

    @Override
    protected void channelRead0(ChannelHandlerContext ctx, ClientPacket packet) {
        var packetId = packet.type().id();
        var handler = handlers[packetId];

        if (handler == null) {
            throw new IllegalStateException("Packet " + packet.type() + " has no attached handler");
        }
        var buffer = packet.buffer();

        try {
            handler.handle(gameServer, player, buffer);
        } finally {
            buffer.release();
        }
    }

    @Override
    public void exceptionCaught(ChannelHandlerContext ctx, Throwable cause) {
        LOGGER.error("Exception caught in dispatcher", cause);
        ctx.disconnect();
    }
}

File: game-server/src/main/kotlin/work/fking/pangya/game/net/ClientPacketDispatcher.kt
package work.fking.pangya.game.net

import io.netty.channel.ChannelHandlerContext
import io.netty.channel.SimpleChannelInboundHandler
import io.netty.channel.unix.Errors
import org.slf4j.LoggerFactory
import work.fking.pangya.game.GameServer
import work.fking.pangya.game.player.Player

private val LOGGER = LoggerFactory.getLogger(ClientPacketDispatcher::class.java)

class ClientPacketDispatcher(
    private val gameServer: GameServer,
    private val player: Player,
    private val handlers: Array<ClientPacketHandler?>
) : SimpleChannelInboundHandler<ClientPacket>() {

    override fun channelRead0(ctx: ChannelHandlerContext, packet: ClientPacket) {
        val packetId = packet.type.id()
        val handler = handlers[packetId] ?: throw IllegalStateException("Packet " + packet.type + " has no attached handler")
        val buffer = packet.buffer
        try {
            handler.handle(gameServer, player, buffer)
        } finally {
            buffer.release()
        }
    }

    override fun exceptionCaught(ctx: ChannelHandlerContext, cause: Throwable) {
        when (cause) {
            is Errors.NativeIoException -> ctx.disconnect()
            else -> {
                LOGGER.error("Exception caught in dispatcher", cause)
                ctx.disconnect()
            }
        }
    }
}
